<a href="https://colab.research.google.com/github/zfukuoka/Training-KikaiGakusyuZukan/blob/master/Chap1_2_testSupervizedML_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1.2 機械学習に必要なステップ

## 教師あり学習（分類）の例

ポイントは sklearn を用いてアメリカウィスコンシン州の乳がんデータを用いて教師あり学習（ロジスティック回帰）を動かすこと。

以下、テキストに沿って幾つかのセルで分割して記載。

* 変数 data にデータセットの読み込み

In [1]:
from sklearn.datasets import load_breast_cancer
data = load_breast_cancer()


* 特徴量を 変数Xに格納
* 目的変数を 変数yに格納

In [2]:
X = data.data
y = data.target

In [3]:
## テキストにはないが、コードを理解するための補足用出力
# Xとyの型確認
print(type(X))
print(type(y))

# Xとyのレコード数とフィールド数を確認
print(X.shape)
print(y.shape)

# yには悪性と良性の振り分けがあるはずなので、各々の値とレコード数を確認
print(y.max())
print(y.min())
print(y[y==1].shape)
print(y[y==0].shape)

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>
(569, 30)
(569,)
1
0
(357,)
(212,)


* 変数X には全体として569レコード（悪性212レコード、良性357レコード）あり、1レコードに30のデータがある
  * 30のデータは10データ毎に、平均値、エラー値、最悪値に分かれている
* 変数y には変数Xのレコード番号に対応した良性、悪性の結果が入っている

* 今回は平均値のデータを用いて、学習するため、下記のように抽出


In [4]:
X = X[:, :10]

In [5]:
## テキストにはないが、コードを理解するための補足用出力
# Xが30フィールドから10フィールドに減らされていることを確認
print(X.shape)

(569, 10)


## 実装方法

ここで用いるロジスティック回帰は第2章の図鑑番号３：ロジスティック回帰を参照のこと

* scikit-learn の LogisticRegression クラスをインポートして、インスタンスを生成して変数 model に初期化したモデルを代入


In [6]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

* fit() を呼び出し、変数model が学習済みモデルになる

In [7]:
model.fit(X, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

* 下記は学習モデルで予測して、結果を 変数y_pred に代入を行っている
  * **注意：本来は学習データである変数X を使って予測することのは誤り**

In [8]:
y_pred = model.predict(X)

## 評価方法

* scikit-learn の accuracy_score() を用いると、正解率を確認できる
  * 正解率は第4章を参照のこと（たしか、JDLA のG検定でもでてきたところ）



In [9]:
from sklearn.metrics import accuracy_score
accuracy_score(y, y_pred)

0.9121265377855887